### DISCUSSION TITLE
Development moves back to pyscripter

### DISCUSSION THREAD

**Discussion by pyscripter** [TurboPack/SynEdit#239] [Announcements]
Embarcadero expressed their concern about backward incompatibilities introduced in SynEdit, even if they amount to just opening a form ignoring errors and saving the form back.

My priorities are different.  I am striving to make SynEdit the best code editing component available on any language, in the same class with say Visual Studio Code or Scintilla.   To that end:

1. Long ago, I have implemented code folding.
2. GDI painting, character measurement and cursor positioning were replaced with DirectWrite.
3. [Unicode support](https://en.delphipraxis.net/topic/6277-synedit-just-got-a-major-uplift/) is now on a par with the best editors around (surrogate pair and emoji support, bi-directional text editing etc.).
4. [The undo/redo system was buggy and a mess](https://github.com/TurboPack/SynEdit/issues/78), getting in the way of implementing new features.  I has been [reimplemented from scratch](https://github.com/TurboPack/SynEdit/issues/78).
5. The gutter has been [reimplemented from scratch](https://github.com/TurboPack/SynEdit/discussions/86) and is now flexible and extensible.
6. A [track changes bar](https://github.com/TurboPack/SynEdit/discussions/122) like in Visual Studio has been added and unlike Delphi's it saves and restores line state correctly on undo/redo.
7. The code base has been refactored cleaned-up, and partially documented, yet, and despite of the new features, it is thousands of lines shorter than the original. 
8. [Font ligatures are supported](https://github.com/TurboPack/SynEdit/issues/35).
9. New, Word-like, [spell checking component added](https://github.com/TurboPack/SynEdit/issues/237).
10. [Scintilla-like indicators](https://github.com/TurboPack/SynEdit/discussions/160) have been implemented.
11. Performance has massively improved.  See for instance [here](https://en.delphipraxis.net/topic/6277-synedit-just-got-a-major-uplift/?do=findComment&comment=55407).
12. See [here](https://github.com/TurboPack/SynEdit/issues?q=label%3Aenhancement+) for the full list of enhancements in the the TurboPack fork.   Also see the Delphi Praxis posts: [Post 1](https://en.delphipraxis.net/topic/3028-synedit-preferred-version/?do=findComment&comment=24291), [Post 2](https://en.delphipraxis.net/topic/3528-turbo-synedit-font-ligatures/), [Post 3](https://en.delphipraxis.net/topic/6277-synedit-just-got-a-major-uplift/)

The next big feature in the plan is multi-cursor editing like in VS Code. 

Since our priorities differ, I have moved Synedit development back to the pyscripter Github repos.  My work on SynEdit will carry on https://github.com/pyscripter/SynEdit.

**Comment by vincentparrett** [comment]
I have to ask, why do Embarcadero get to decide what happens in this repo - I have yet to see one commit (correct me if I am wrong) from an embarcadero employee. It would seem to me that they do not understand how open source works.

**Reply by vincentparrett** [reply]
It also bugs me that we will now have 2 unmaintained "official" synedit repositories.

**Reply by pyscripter** [reply]
> I have to ask, why do Embarcadero get to decide what happens in this repo

Embarcadero wanted to add value to their product by giving easy access to their customers to a range of widely used components.  The TurboPack organization was set up by them for this purpose.  The Turbopack repositories are used for the GetIt packages.   If some clients upgrade Delphi, get the packages they need from GetIt, and find that their applications does not work out-of-the-box, they may well complain and this is an issue for Embarcadero.  I can understand that.  Backward compatibility is very important for them.  Improving the components was not a priority for them.

**Reply by jimmckeeth** [reply]
What if we fork SynEdit into two versions. Have all active development happen in TSynEdit2....

**Reply by pyscripter** [reply]
> What if we fork SynEdit into two versions. Have all active development happen in TSynEdit2....

Isn't this what is happening now?  New features are implemented in the https://github.com/pyscripter/SynEdit fork.

**Reply by checkdigits** [reply]
> 
> Isn't this what is happening now? New features are implemented in the https://github.com/pyscripter/SynEdit fork.

This is the right option IMO.

**Reply by jimmckeeth** [reply]
I was thinking we could add SynEdit2 to GetIt (this one), although maybe that isn't necessary.
